### ISSUE TITLE
[Documentation] CellProfiler as Bioconda Recipe?

### ISSUE THREAD

**Issue by gwaygenomics** [CellProfiler/CellProfiler#3730]
reading through the conda install [docs](https://github.com/CellProfiler/CellProfiler/wiki/Conda-Installation), I am wondering if anyone has tried putting the package directly on [bioconda](https://github.com/bioconda/bioconda-recipes)? Perhaps this could make install in conda workflows easier.

I see that [prokaryote](https://github.com/CellProfiler/prokaryote) is already [a recipe there](https://github.com/bioconda/bioconda-recipes/tree/3642c20fb306ffcddee91505ddcddec59874748f/recipes/prokaryote) so ensuring proper dependencies may not be too difficult 🤞.

**Comment by bethac07** [comment]
It definitely has not been tried; is this something you have experience doing and might be able to pitch in on? 

@0x00b1 do you have thoughts here?

**Comment by gwaygenomics** [comment]
I have contributed a couple recipes and difficulty typically ranges from 5 minutes to nearly impossible. Based on #3711 it could be difficult, but also maybe worthwhile.

Would it be useful to have a direct conda install? Would it make things easier for developers (current or new)?

**Comment by bethac07** [comment]
I don't have a good sense of the overlap between "prefers conda to just managing their own packages" and "is able/willing to contribute to a code base this fiddly"; I suspect it's neither 0% nor 100% but no idea beyond that.

**Comment by gwaygenomics** [comment]
update: CellProfiler was added to bioconda on March 20th, 2020. Recipe here: https://github.com/bioconda/bioconda-recipes/tree/16bd51581b66ee7f5ed35b545f368fc3740a9662/recipes/cellprofiler

https://bioconda.github.io/recipes/cellprofiler/README.html

**Comment by gwaygenomics** [comment]
@mfractal18 is thinking about updating the bioconda recipe to CellProfiler version 4. The goal of the update is to be able to run CellProfiler 4 in [Elyra](https://github.com/elyra-ai/elyra).

Updating the bioconda recipe may be a more stable and permanent solution than those provided in #3833, #3838, and #3902

## How to update

It looks like updating recipes in bioconda ([link to guide here](https://bioconda.github.io/contributor/updating.html)) has gotten easier with [`bioconda-utils`](https://bioconda.github.io/recipes/bioconda-utils/README.html#package-bioconda-utils).

## Important considerations

In addition to CellProfiler, of the [other recipes](https://github.com/bioconda/bioconda-recipes/search?q=cellprofiler) it looks like only `python-bioformats` needs updating. Although, I don't think this is necessary since python-bioformats version 4.0 can be installed via pip.

| Package | Current Bioconda Version | CellProfiler 4 Required Version |
| :-------: | :-------------------------: | :-----------------------------: |
| [`centrosome`](https://github.com/bioconda/bioconda-recipes/blob/4d8723c5b52dd4310934da589de887198784e7b6/recipes/centrosome/meta.yaml) | 1.2.0 | 1.2.0 |
| [`prokaryote`](https://github.com/bioconda/bioconda-recipes/blob/2ed27c3d4d3101cbc05e41c66f4153022c9af039/recipes/prokaryote/meta.yaml) | 2.4.2 | 2.4.2 |
| [`python-bioformats`](https://github.com/bioconda/bioconda-recipes/blob/448b7d9aa7328a80f744ee25e017dfeb8eef09cd/recipes/python-bioformats/meta.yaml) | 1.5.2 | 4.0.0 |

**Comment by minhtran1309** [comment]
Hi @gwaygenomics, 
Thanks for creating conda recipe.  May I ask when CellProfiler v4.0 be available in bioconda? I am running MacOS 10.14 and v4.0 doesn't support my Os. I am thinking to install it from the source but it's pretty problematic as well. 

Looking forward to hearing your conda update. 
Thanks

**Comment by raivivek** [comment]
Hi all, I've been wanting to install CellProfiler through Conda as well. I'm working on resolving some dependencies and just submitted a PR to update the python-bioformats recipe. Hopefully, will get it done soon. Please feel free to hop-in and correct my mistakes.

**Comment by raivivek** [comment]
@gwaygenomics The issue I'm facing is this. It appears that there are two sources of python-javabridge on PyPI. One that says, `python-javabridge` and other says `javabridge`. However, looks like both point to the same GitHub source repository. Which one should be the correct choice here?

**Comment by gwaygenomics** [comment]
@minhtran1309 

> Thanks for creating conda recipe. May I ask when CellProfiler v4.0 be available in bioconda? I am running MacOS 10.14 and v4.0 doesn't support my Os. I am thinking to install it from the source but it's pretty problematic as well.

It was actually @sunyi000 who added CellProfiler 3.1.9 to bioconda in bioconda/bioconda-recipes#20780.

Also, we don't know when CellProfiler 4.0 will be available in bioconda. As an open source repository, anyone can contribute to bioconda. This means that the contribution (which is apparently in high demand!) is up for grabs. @mfractal18 was considering performing the update, but if you (or anyone else) is able to upgrade sooner, we are supportive!

@raivivek 

> Hi all, I've been wanting to install CellProfiler through Conda as well. I'm working on resolving some dependencies and just submitted a PR to update the python-bioformats recipe. Hopefully, will get it done soon. Please feel free to hop-in and correct my mistakes.

Great! I am adding a reference here for posterity bioconda/bioconda-recipes#24891

Also, the bioconda maintainers are fantastic, and will help you to get the recipe pass the tests. 

> The issue I'm facing is this. It appears that there are two sources of python-javabridge on PyPI. One that says, python-javabridge and other says javabridge. However, looks like both point to the same GitHub source repository. Which one should be the correct choice here?

hmm, I am not sure (I have very little familiarity with the code-base). Have you looked for the answer in https://github.com/LeeKamentsky/python-javabridge? The issues are a good place to start the search.

Thanks for your interest in the project, and for your contributions

**Comment by gwaygenomics** [comment]
note that @raivivek has filed bioconda/bioconda-recipes#24967 and has merged in bioconda/bioconda-recipes#24891 updating python-bioformats. Thanks @raivivek !

**Comment by gwaygenomics** [comment]
## Progress update

* @raivivek merged in `cellprofiler-core` (version 4.0.6) in bioconda/bioconda-recipes#24967
* @raivivek filed a PR to update `CellProfiler` (version 4.0.6) in bioconda/bioconda-recipes#24983

For the CellProfiler bioconda PR, we are stuck because of a `wxpython` version issue:

in https://app.circleci.com/pipelines/github/bioconda/bioconda-recipes/34792/workflows/8dde8898-8db5-43fe-b5cd-acc1282082ff/jobs/128233:

<details>
  <summary>Click to expand error</summary>

```bash
20:47:58 BIOCONDA INFO (OUT) Traceback (most recent call last):
20:47:58 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda_build/environ.py", line 764, in get_install_actions
20:47:58 BIOCONDA INFO (OUT)     actions = install_actions(prefix, index, specs, force=True)
20:47:58 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda/common/io.py", line 88, in decorated
20:47:58 BIOCONDA INFO (OUT)     return f(*args, **kwds)
20:47:58 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda/plan.py", line 474, in install_actions
20:47:58 BIOCONDA INFO (OUT)     txn = solver.solve_for_transaction(prune=prune, ignore_pinned=not pinned)
20:47:58 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda/core/solve.py", line 117, in solve_for_transaction
20:47:58 BIOCONDA INFO (OUT)     should_retry_solve)
20:47:58 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda/core/solve.py", line 158, in solve_for_diff
20:47:58 BIOCONDA INFO (OUT)     force_remove, should_retry_solve)
20:47:58 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda/core/solve.py", line 275, in solve_final_state
20:47:58 BIOCONDA INFO (OUT)     ssc = self._add_specs(ssc)
20:47:58 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda/core/solve.py", line 565, in _add_specs
20:47:58 BIOCONDA INFO (OUT)     explicit_pool = ssc.r._get_package_pool(self.specs_to_add)
20:47:58 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda/resolve.py", line 556, in _get_package_pool
20:47:58 BIOCONDA INFO (OUT)     pool = self.get_reduced_index(specs)
20:47:58 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda/common/io.py", line 88, in decorated
20:47:58 BIOCONDA INFO (OUT)     return f(*args, **kwds)
20:47:58 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda/resolve.py", line 577, in get_reduced_index
20:47:58 BIOCONDA INFO (OUT)     explicit_specs, features = self.verify_specs(explicit_specs)
20:47:58 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda/resolve.py", line 288, in verify_specs
20:47:58 BIOCONDA INFO (OUT)     raise ResolvePackageNotFound(bad_deps)
20:47:58 BIOCONDA INFO (OUT) conda.exceptions.ResolvePackageNotFound:
20:47:58 BIOCONDA INFO (OUT)   - wxpython[version='>=4.1.0']
20:47:58 BIOCONDA INFO (OUT) 
20:47:58 BIOCONDA INFO (OUT) During handling of the above exception, another exception occurred:
20:47:58 BIOCONDA INFO (OUT) 
20:47:58 BIOCONDA INFO (OUT) Traceback (most recent call last):
20:47:58 BIOCONDA INFO (OUT)   File "/opt/conda/bin/conda-build", line 11, in <module>
20:47:58 BIOCONDA INFO (OUT)     sys.exit(main())
20:47:58 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda_build/cli/main_build.py", line 474, in main
20:47:58 BIOCONDA INFO (OUT)     execute(sys.argv[1:])
20:47:58 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda_build/cli/main_build.py", line 465, in execute
20:47:58 BIOCONDA INFO (OUT)     verify=args.verify, variants=args.variants)
20:47:58 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda_build/api.py", line 209, in build
20:47:58 BIOCONDA INFO (OUT)     notest=notest, need_source_download=need_source_download, variants=variants)
20:47:58 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda_build/build.py", line 2864, in build_tree
20:47:58 BIOCONDA INFO (OUT)     notest=notest,
20:47:58 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda_build/build.py", line 1911, in build
20:47:58 BIOCONDA INFO (OUT)     create_build_envs(top_level_pkg, notest)
20:47:58 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda_build/build.py", line 1795, in create_build_envs
20:47:58 BIOCONDA INFO (OUT)     raise e
20:47:58 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda_build/build.py", line 1785, in create_build_envs
20:47:58 BIOCONDA INFO (OUT)     channel_urls=tuple(m.config.channel_urls))
20:47:58 BIOCONDA INFO (OUT)   File "/opt/conda/lib/python3.7/site-packages/conda_build/environ.py", line 766, in get_install_actions
20:47:58 BIOCONDA INFO (OUT)     raise DependencyNeedsBuildingError(exc, subdir=subdir)
20:47:58 BIOCONDA INFO (OUT) conda_build.exceptions.DependencyNeedsBuildingError: Unsatisfiable dependencies for platform linux-64: {"wxpython[version='>=4.1.0']"}
20:48:02 BIOCONDA ERROR COMMAND FAILED (exited with 1): docker run -t --net host --rm -v /tmp/tmpx0klnkra/build_script.bash:/opt/build_script.bash -v /home/circleci/project/miniconda/conda-bld/:/opt/host-conda-bld -v /home/circleci/project/recipes/cellprofiler:/opt/recipe -e LANG=en_US.UTF-8 -e HOST_USER_ID=1001 bioconda/bioconda-utils-build-env:latest /bin/bash /opt/build_script.bash

20:48:02 BIOCONDA ERROR BUILD FAILED recipes/cellprofiler
20:48:02 BIOCONDA INFO (COMMAND) conda build purge
20:48:02 BIOCONDA ERROR BUILD SUMMARY: of 1 recipes, 1 failed and 0 were skipped. Details of recipes and environments follow.
20:48:02 BIOCONDA ERROR BUILD SUMMARY: FAILED recipe recipes/cellprofiler
```
</details>

The conda-forge folks decided to merge the `wxpython` build even though the windows build is failing (https://github.com/conda-forge/wxpython-feedstock/pull/48#issuecomment-713480757). However, a more recent error (`ld: library not found for -lwx_osx_cocoau_core-3.1`) in the Mac build is preventing us from merging (https://github.com/conda-forge/wxpython-feedstock/pull/48#issuecomment-713514411)

https://github.com/conda-forge/wxpython-feedstock/pull/48#issuecomment-713514411

**Comment by gwaygenomics** [comment]
Note CP 4.2 is now on bioconda - https://github.com/bioconda/bioconda-recipes/commit/9223415e488b2c226eeb9892acfb47042cd1ce4c#diff-70bad441c4f9ed0fa3cd1b7fa20b92e26cd9395f2efc4ac0b6013bc8b76f2662
